### PR TITLE
add `willClose` to hs.webview navigation callback

### PR DESCRIPTION
**Pull request by asmagill**
The navigation callback function will now receive a `willClose` message when the window is closing.

Addresses #1259

### CODE REVIEW

**Comment by latenitefilms** [comment]
Legend, thank you!

**Comment by cmsj** [comment]
lgtm, merge at will :)

**Comment by asmagill** [comment]
@latenitefilms, note that there is a change in syntax (new callback method) since the newer methods didn't fit quite as nicely into navigationCallback:
~~~
hs.webview:windowCallback(fn) -> webviewObject
Method
Set or clear a callback for updates to the webview window

Parameters:
 * `fn` - the function to be called when the webview window is moved or closed. Specify an explicit nil to clear the current callback.  The function should expect 2 or 3 arguments and return none.  The arguments will be one of the following:

   * "closing", webview - specifies that the webview window is being closed, either by the user or with the [hs.webview:hide](#hide) method.
     * `action`  - in this case "closing", specifying that the webview window is being closed
     * `webview` - the webview that is being closed

   * "focusChange", webview, state - indicates that the webview window has either become or stopped being the focused window
     * `action`  - in this case "focusChange", specifying that the webview window is being closed
     * `webview` - the webview that is being closed
     * `state`   - a boolean, true if the webview has become the focused window, or false if it has lost focus

   * "frameChange", webview, frame - indicates that the webview window has been moved or resized
     * `action`  - in this case "focusChange", specifying that the webview window is being closed
     * `webview` - the webview that is being closed
     * `frame`   - a rect-table containing the new co-ordinates and size of the webview window

Returns:
 * The webview object
~~~

The callback may get multiple "frameChange" messages depending upon how quickly you resize/move the window... this is determined by how often the OS decides to notify us and at present I don't think I can work around that at the obj-c level, so depending upon what you're doing based upon these changes, you may want to stick a timer in there to make sure that no new updates come within some timeout you choose before performing the desired action.

**Comment by latenitefilms** [comment]
AMAZING! Thanks so much - HUGELY appreciated @asmagill !

Will test this out properly over the weekend, and let you know how we go.

Once this is merged into `master`, I wonder if it's worth updating `hs.doc.hsdocs` to use this new API?

**Comment by asmagill** [comment]
@latenitefilms, probably a good idea... I'll address that when I decide how to handle the toolbar issue I uncovered and just added to the 0.9.53 todo list.

**Comment by latenitefilms** [comment]
@asmagill - Thanks so much for all your amazing work on this - HUGELY appreciated!

Is there a way to differentiate between when a user clicks the "close" button on a `hs.webview` versus when the code triggers a `hs.webview:hide()` command?

Basically, I'm trying to use a `hs.webview` as a HUD for a specific application, so I need a good and reliable way of ONLY showing the HUD when that application is active, hidden when it's not, and disabling the HUD if the user closes it (as opposed to my code which will "hide" it when the specific application is no longer visible).

Thoughts?

**Comment by latenitefilms** [comment]
Also, FYI - I'm not using a Hammerspoon Dock icon, which makes this whole process a bit tricky, as it's really hard to work out if the `hs.webview` HUD should be hidden or not when dragging, or when "Keep console window on top" is active.

**Comment by latenitefilms** [comment]
Actually... upon further testing, it doesn't seem like the callback is triggered when `hs.webview:hide()` is called anyway, which is great for me! Maybe the documentation needs to be tweaked?

**Comment by asmagill** [comment]
Hmm... It's supposed to be, unless I'm missing something... I'll test it further tomorrow and see if I can figure out why and update the docs/code accordingly.

**Comment by latenitefilms** [comment]
Also, I noticed that the `closing` callback was being triggered when Hammerspoon is quit or reloaded - which I didn't really want (really I just wanted to know when the user was clicking the "close" button). I've come up with a way to ignore this using `hs.shutdownCallback`, which works fine, but I wonder if it's worth adding an additional option so that the callback itself can let you know what's actually causing the closing callback?

**Comment by latenitefilms** [comment]
Awesome! All looks good to me!

Thanks again for all your help!!

**Comment by asmagill** [comment]
Just noticed your comment about being invoked on reload/restart of Hammerspoon... this is because the window callback is removed *after* the garbage collection function closes the window... this was the simplest way to ensure that the `hs.webview:delete` does invoked the callback.

It wouldn't be that hard to change, though, so that delete explicitly invokes the callback.  Thoughts?

**Comment by asmagill** [comment]
I guess I'm leaning towards making the change... nothing else triggers a callback during garbage collection.

**Comment by latenitefilms** [comment]
Legend, thanks @asmagill ! The only thing I really care about is when the end-user clicks the "close" button of the `hs.webview`. I don't really care if the callback is trigged by `hs.webview:delete`, because if I'm the one triggering `hs.webview:delete` anyway - then I know it's being triggered, and can add code to do whatever I need to.

Will test out your code properly soon - but looks good to me!

**Comment by latenitefilms** [comment]
@asmagill  - Ummm... your latest changes seem to stop the callback from being triggered when the close button is pressed?

**Comment by asmagill** [comment]
Using the built in hsdocs browser:

~~~lua
> hs.doc.hsdocs.help("hs")
> hs.doc.hsdocs._browser:windowCallback(function(...) print(timestamp(), finspect(table.pack(...))) end)
hs.webview: Hammerspoon docs: hs module (0x6080012520e8)
2017-04-12 01:05:46.3495	{ "focusChange", <userdata 1> -- hs.webview: Hammerspoon docs: hs module (0x6080012520e8), true, n = 3 }
2017-04-12 01:05:51.0769	{ "closing", <userdata 1> -- hs.webview: Hammerspoon docs: hs module (0x6080012520e8), n = 2 }
2017-04-12 01:05:51.0801	{ "focusChange", <userdata 1> -- hs.webview: Hammerspoon docs: hs module (0x6080012520e8), false, n = 3 }

> hs.doc.hsdocs.help("hs")
2017-04-12 01:06:05.4711	{ "focusChange", <userdata 1> -- hs.webview: Hammerspoon docs: hs module (0x6080012520e8), true, n = 3 }

2017-04-12 01:06:07.7895	{ "closing", <userdata 1> -- hs.webview: Hammerspoon docs: hs module (0x6080012520e8), n = 2 }
2017-04-12 01:06:07.793 	{ "focusChange", <userdata 1> -- hs.webview: Hammerspoon docs: hs module (0x6080012520e8), false, n = 3 }
~~~

The first "closing" line was by clicking on the red close button for the browser window, the second from using Cmd-W...

**Comment by latenitefilms** [comment]
Ummm... that's weird. I'll do some more testing, but [534500b](https://github.com/Hammerspoon/hammerspoon/pull/1299/commits/534500b8b69fe7b2d486b5aa8afb09070d8e2654) seemed to break things for me. Reverting back to the previous commit seemed to fix everything.

**Comment by asmagill** [comment]
Do you set `hs.webview:deleteOnClose(true)` or creating additional windows through javascript or frame targets?  It just crossed my mind that this might not be caught in such a case because `delete` is not bing explicitly called in such a case.  I'll take a look at it in a bit...

**Comment by latenitefilms** [comment]
Not sure why but [f03ff74](https://github.com/Hammerspoon/hammerspoon/pull/1299/commits/f03ff74ba929fe821a3fa5afcda40f87da2bb325) works fine for me - [534500b](https://github.com/Hammerspoon/hammerspoon/pull/1299/commits/534500b8b69fe7b2d486b5aa8afb09070d8e2654) does not.

Yes, `deleteOnClose` is `true`. There's only one window.

`hs.webview` code as follows:

```
		hud.webview = webview.new(getHUDRect(), developerExtrasEnabled, hud.webviewController)
			:windowStyle({"HUD", "utility", "titled", "nonactivating", "closable", "resizable"})
			:shadow(true)
			:closeOnEscape(true)
			:html(hud.generateHTML())
			:allowGestures(false)
			:allowNewWindows(false)
			:windowTitle(hud.title)
			:level(drawing.windowLevels.floating)
			:windowCallback(windowCallback)
			:deleteOnClose(true)
```

**Comment by asmagill** [comment]
Ok, I see the problem... give me a sec...

**Comment by latenitefilms** [comment]
@asmagill - Unfortunately [cef368d](https://github.com/Hammerspoon/hammerspoon/pull/1299/commits/cef368d393670b6f31b1b85af2591fc7c1f00804) seems to close Hammerspoon all together when I close a `hs.webview` now.

**Comment by asmagill** [comment]
@latenitefilms I'm not sure what you mean by close Hammerspoon all together... do you mean that it's causing the Hammerspoon application to quit?

**Comment by latenitefilms** [comment]
Yes, when I closed the `hs.webview`, the whole Hammerspoon application quit.

**Comment by asmagill** [comment]
I can't replicate that behavior with

~~~lua
> webview = hs.webview.new({x=100,y=100,h=500,w=500}, {developerExtrasEnabled = true}, hs.webview.usercontent.new("hud"))
			:windowStyle({"HUD", "utility", "titled", "nonactivating", "closable", "resizable"})
			:shadow(true)
			:closeOnEscape(true)
			:html([[<h1>BooYah!</h1>]])
			:allowGestures(false)
			:allowNewWindows(false)
			:windowTitle("A window")
			:level(hs.drawing.windowLevels.floating)
			:windowCallback(function(...) print(timestamp(), finspect(table.pack(...))) end)
			:deleteOnClose(true):show()
-- Loading extension: webview
-- Loading extension: drawing

2017-04-12 10:24:39.7902	{ "closing", <userdata 1> -- hs.webview: (0x60000065c6d8), n = 2 }
~~~

I'll try installing https://github.com/CommandPost/CommandPost later today and see what might be different then just this modified snippit

It's odd that this latest pull should be causing problems... strictly speaking, the original code was "wrong" in that `hs.webview` was forcing the window closed (when deleteOnClose was true) during the delegate call to `windowShallClose:` when all it should have been doing was reporting back whether or not to allow the close...

**Comment by latenitefilms** [comment]
Let me try again, maybe I was doing something funky.

**Comment by latenitefilms** [comment]
I just re-downloaded the latest Hammerspoon code, and tried this branch, and it seemed to work as expected, using your example code above.

However, for some reason it's causing my Hammerspoon fork to crash:

```
2017-04-13 01:59:16.276885 CommandPost[10199:1164941] logForLuaSkin: error, state ERRRUN: C stack overflow
2017-04-13 01:59:16.283918 CommandPost[10199:1164941] logForLuaSkin: error, state ERRRUN: C stack overflow
2017-04-13 01:59:16.284253 CommandPost[10199:1164941] logForLuaSkin: error, state ERRRUN: C stack overflow
2017-04-13 01:59:16.378248 CommandPost[10199:1164941] *** -[HSWebViewWindow release]: message sent to deallocated instance 0x608495481ee0
```

No idea why (as we shouldn't/haven't have messed with any of the core Hammerspoon code), but the good news is, it seems to be a CommandPost bug, not a Hammerspoon bug!

**Comment by latenitefilms** [comment]
Here's the first section of the crash log if it's of interest?

```
#0. Crashed: com.apple.main-thread
0   AppKit                         0x7fffd309fa2c -[NSWindow release] + 195
1   Foundation                     0x7fffd6f72bc2 -[NSConcreteNotification dealloc] + 64
2   AppKit                         0x7fffd3323548 __18-[NSWindow _close]_block_invoke + 205
3   AppKit                         0x7fffd332342c -[NSWindow _close] + 365
4   internal.so                    0x106f42b32 userdata_gc + 158
5   LuaSkin                        0x10049d897 luaD_precall (ldo.c:454)
6   LuaSkin                        0x10049d9dd luaD_call (ldo.c:515)
7   LuaSkin                        0x10049da2a luaD_callnoyield (ldo.c:527)
8   LuaSkin                        0x1004a565f luai_objcttry (lobjectivec_exceptions.m:173)
9   LuaSkin                        0x10049ded6 luaD_pcall (ldo.c:162)
10  LuaSkin                        0x1004ae62c lua_pcallk (lapi.c:968)
11  internal.so                    0x106f42a11 -[HSWebViewWindow windowWillClose:] + 338
12  CoreFoundation                 0x7fffd55719fc __CFNOTIFICATIONCENTER_IS_CALLING_OUT_TO_AN_OBSERVER__ + 12
13  CoreFoundation                 0x7fffd55718fb _CFXRegistrationPost + 427
14  CoreFoundation                 0x7fffd5571662 ___CFXNotificationPost_block_invoke + 50
15  CoreFoundation                 0x7fffd552fd12 -[_CFXNotificationRegistrar find:object:observer:enumerator:] + 2018
16  CoreFoundation                 0x7fffd552ecfb _CFXNotificationPost + 667
17  Foundation                     0x7fffd6f72a37 -[NSNotificationCenter postNotificationName:object:userInfo:] + 66
18  AppKit                         0x7fffd3323548 __18-[NSWindow _close]_block_invoke + 205
19  AppKit                         0x7fffd332342c -[NSWindow _close] + 365
20  internal.so                    0x106f42b32 userdata_gc + 158
21  LuaSkin                        0x10049d897 luaD_precall (ldo.c:454)
22  LuaSkin                        0x10049d9dd luaD_call (ldo.c:515)
23  LuaSkin                        0x10049da2a luaD_callnoyield (ldo.c:527)
24  LuaSkin                        0x1004a565f luai_objcttry (lobjectivec_exceptions.m:173)
25  LuaSkin                        0x10049ded6 luaD_pcall (ldo.c:162)
26  LuaSkin                        0x1004ae62c lua_pcallk (lapi.c:968)
27  internal.so                    0x106f42a11 -[HSWebViewWindow windowWillClose:] + 338
28  CoreFoundation                 0x7fffd55719fc __CFNOTIFICATIONCENTER_IS_CALLING_OUT_TO_AN_OBSERVER__ + 12
29  CoreFoundation                 0x7fffd55718fb _CFXRegistrationPost + 427
30  CoreFoundation                 0x7fffd5571662 ___CFXNotificationPost_block_invoke + 50
31  CoreFoundation                 0x7fffd552fd12 -[_CFXNotificationRegistrar find:object:observer:enumerator:] + 2018
32  CoreFoundation                 0x7fffd552ecfb _CFXNotificationPost + 667
33  Foundation                     0x7fffd6f72a37 -[NSNotificationCenter postNotificationName:object:userInfo:] + 66
34  AppKit                         0x7fffd3323548 __18-[NSWindow _close]_block_invoke + 205
35  AppKit                         0x7fffd332342c -[NSWindow _close] + 365
36  internal.so                    0x106f42b32 userdata_gc + 158
37  LuaSkin                        0x10049d897 luaD_precall (ldo.c:454)
38  LuaSkin                        0x10049d9dd luaD_call (ldo.c:515)
39  LuaSkin                        0x10049da2a luaD_callnoyield (ldo.c:527)
40  LuaSkin                        0x1004a565f luai_objcttry (lobjectivec_exceptions.m:173)
41  LuaSkin                        0x10049ded6 luaD_pcall (ldo.c:162)
42  LuaSkin                        0x1004ae62c lua_pcallk (lapi.c:968)
43  internal.so                    0x106f42a11 -[HSWebViewWindow windowWillClose:] + 338
44  CoreFoundation                 0x7fffd55719fc __CFNOTIFICATIONCENTER_IS_CALLING_OUT_TO_AN_OBSERVER__ + 12
45  CoreFoundation                 0x7fffd55718fb _CFXRegistrationPost + 427
46  CoreFoundation                 0x7fffd5571662 ___CFXNotificationPost_block_invoke + 50
47  CoreFoundation                 0x7fffd552fd12 -[_CFXNotificationRegistrar find:object:observer:enumerator:] + 2018
48  CoreFoundation                 0x7fffd552ecfb _CFXNotificationPost + 667
49  Foundation                     0x7fffd6f72a37 -[NSNotificationCenter postNotificationName:object:userInfo:] + 66
50  AppKit                         0x7fffd3323548 __18-[NSWindow _close]_block_invoke + 205
51  AppKit                         0x7fffd332342c -[NSWindow _close] + 365
```

**Comment by latenitefilms** [comment]
Note to self:

```
<Ng> that means that something is causing the HSWebViewWindow object to be deleted twice
<Ng> things like that can happen if the binding of the object between Lua and ObjC isn't tight enough, or if the object gets accidentally garbage collected by Lua before an explicit release call, or if there's a logic bug in something like the delete-on-close window thing
```

**Comment by asmagill** [comment]
Ok, I think I see the problem, but I'm not entirely sure the solution yet... line 20 and line 4 show that the object is getting released twice, likely because of the close invoked in the garbage collection... this is a "just in case" because it really should be closed already before it ever reaches this line, and in any case, the callback and metatables are already gone, so... I'm not entirely sure why it should matter, but let me try something...

**Comment by asmagill** [comment]
let me know if this makes a difference

**Comment by latenitefilms** [comment]
Victory! Yep - that works great - thank you!

**Comment by asmagill** [comment]
@cmsj objections to merging this and #1317?  I'll be fixing the toolbar documentation and looking into spoon documentation tonight, so closing these two won't quite get us to 0.9.53, but very close IIRC.

**Comment by cmsj** [comment]
@asmagill no objections at all :)

**Comment by latenitefilms** [comment]
@asmagill - Sorry to be a pain, but I think the callback is still being triggered if you have a `hs.webview` open when you quit Hammerspoon?

**Comment by asmagill** [comment]
@latenitefilms can you provide a code snippit (or link) to an example of this?  I can't replicate it myself -- during a reload or quit, the callback does not seem to be invoked on my machine
